### PR TITLE
Add a command to calculate DRep metadata hashes

### LINKED ISSUE #130: [FEATURE] - command to calculate DRep metadata hash
## What

The DRep registration certificate will include an anchor (url, metadata-hash); therefore we need a command to help users to calculate the dreps metadata hash. It should be something similar to what we already do for stake pools, like: 

```
cardano-cli conway governance drep metadata-hash 
--drep-metadata-file FILE
[--out-file FILE]
```

## Why
[From CIP1694](https://github.com/cardano-foundation/CIPs/blob/e96681908fd336702b9d89827063a190c5656eab/CIP-1694/README.md?plain=1#LL398C45-L398C45)

> DRep registration certificates include:
> 
> * a DRep ID
> * a deposit
> * an anchor
> 
> An **anchor** is a pair of:
> 
> * a URL to a JSON payload of metadata
> * a hash of the contents of the metadata URL

## Personas

- [ ] Constitutional committee members
- [x] Delegate representatives
- [ ] SPOs
- [ ] dApp Devs
- [ ] Exchanges
- [ ] Wallets
- [ ] 3rd party tools
- [ ] ADA holders 

## Acceptance Criteria

1. The cardano-cli conway governance drep metadata-hash command should be available and functional.
2. The command should require the `--drep-metadata-file FILE `option to specify the file containing the DRep metadata in JSON format.
3. The command should support the `--out-file FILE` option (optional) to enable users to save the calculated metadata hash to the specified file.
4. The command should calculate the hash of the contents of the JSON payload provided in the `--drep-metadata-file` and produce the resulting metadata hash.
5. The calculated metadata hash should adhere to the hashing algorithm specified by ledger
7. If the `--out-file FILE ` option is specified, the calculated metadata hash should be saved to the specified file.
8. The feature implementation should be well-documented, providing clear usage instructions for the `cardano-cli conway governance drep metadata-hash` command.
9. The command should handle errors gracefully and provide helpful error messages when required options are missing or invalid inputs are provided.
10. The feature should be thoroughly tested to ensure the correctness of the metadata hash calculation process and its compatibility with the anchor requirements in DRep registration certificates.
12. The documentation should be updated to include the new feature, explaining its purpose and usage.

## Definition of Done (DoD)

- [ ] [Acceptance Criteria + User Stories & DoD](https://input-output.atlassian.net/wiki/spaces/QA/pages/3584229468/Acceptance+Criteria+DoD) created and singed-off (by PO, dev & test owners)
- [ ] Builds successfully on CI
- [ ] Code & Test review (as per Acceptance Criteria)
- [ ] There is documentation and/or examples for the new functionality (usage/response)
- [ ] Log/record changes on **Vnext** (or similar depending on what we adopt)
- [ ] Ticket number(s) included in PR description
- [ ] All Acceptance Criteria met and covered by dev/unit/property/integration tests
- [ ] System/E2E automated tests + System Test Engineer Owner Sign-off

# Sign-off

- [ ] Product Owner
- [ ] Dev Owner
- [ ] System Test Engineer Owner

> **_NOTE:_**  Ideally, we should merge only fully implemented and tested features into the master branch. 
> So all the above steps are required for the PR to be merged.  
> In order to avoid the PRs becoming stale and requiring to be rebased on master, these can be merged 
> after a reasonable time (current agreement is 3 days) if the System Test Engineer Owner's sign-off 
> was not provided (last step in the DoD).

> **_IMPORTANT:_** Any deviation from the plan should be discussed and agreed as a comment in the Feature file. 

## Related PRs

### PR DESCRIPTION
**Pull request by carlhammann**
# Changelog

```yaml
- description: |
    Add a command to calculate DRep metadata hashes
# uncomment types applicable to the change:
  type:
  - feature        # introduces a new feature
  # - breaking       # the API has changed in a breaking way
  - compatible     # the API has changed but is non-breaking
  # - optimisation   # measurable performance improvements
  # - improvement    # QoL changes e.g. refactoring
  # - bugfix         # fixes a defect
  # - test           # fixes/modifies tests
  # - maintenance    # not directly related to the code
  # - release        # related to a new release preparation
  # - documentation  # change in code docs, haddocks...
```

# Context

This closes #130 

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] The change log section in the PR description has been filled in
- [ ] New tests are added if needed and existing tests are updated.  These may include:
  - golden tests
  - property tests
  - round trip tests
  - integration tests
  See [Running tests](https://github.com/input-output-hk/cardano-node-wiki/wiki/Running-tests) for more details
- [ ] The version bounds in `.cabal` files are updated
- [ ] CI passes. See note on CI.  The following CI checks are required:
  - [ ] Code is linted with `hlint`.  See `.github/workflows/check-hlint.yml` to get the `hlint` version
  - [ ] Code is formatted with `stylish-haskell`.  See `.github/workflows/stylish-haskell.yml` to get the `stylish-haskell` version
  - [ ] Code builds on Linux, MacOS and Windows for `ghc-8.10.7` and `ghc-9.2.7`
- [ ] The changelog section in the PR is updated to describe the change
- [ ] Self-reviewed the diff

### CODE REVIEW

File: cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/DRep.hs
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/DRep.hs]: Suggest to revert to previous formatting.

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs]: We have functions like `writeByteStringOutput` that case on the `Maybe` for you.

**Comment by carlhammann** [comment]
I'm unsure about a few things here:
1. What this command does (reading a file and hashing its contents) is technically not specific to an era, and therefore the command is now available in every non-Byron era. Is this semantically correct?
2. We can't test the `GovernanceCmdDRepMetadataValidationError` is redered correctly, at least not without a hack, since the constructor of `DRepMetadataValidationError` is not exported by `cardano-api`, it seems. Does this warrant a PR there?
3. I've not yet written a test here, because it feels rather empty to test hashing a file, and the similar `stake-pool metadata-hash` command isn't tested as well. Should I write tests, or do we consider them an unnecessary burden in this case?

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/DRep.hs
Comment (newhoggy) [inline cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/DRep.hs]: This function should take an `era` like the other command functions.  The `era` argument is used by the function to determine if the command is present in the current era.

File: cardano-cli/test/cardano-cli-golden/files/golden/help.cli
Comment (newhoggy) [inline cardano-cli/test/cardano-cli-golden/files/golden/help.cli]: The `governance drep metadata-hash` command doesn't make sense for eras before Conway.

Doing this should resolve that issue: https://github.com/input-output-hk/cardano-cli/pull/308/files#r1337080474

File: cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/DRep.hs
Comment (carlhammann) [reply cardano-cli/src/Cardano/CLI/EraBased/Options/Governance/DRep.hs]: ~~See my question 1 above. To which eras should I restrict?~~ Answered below, thanks!

File: cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/DRep.hs
Comment (carlhammann) [reply cardano-cli/src/Cardano/CLI/EraBased/Commands/Governance/DRep.hs]: Whoops, yes!

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs
Comment (carlhammann) [reply cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs]: Nice, good to know about that module in `cardano-api`!

**Review by newhoggy — APPROVED** [review]

**Comment by newhoggy** [comment]
Nice!

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs]: Should be:
```
runGovernanceDRepMetadataHashCmd
  ::   DRepMetadataFile In
```

File: cardano-cli/test/cardano-cli-golden/Test/Golden/ErrorsSpec.hs
Comment (Jimbo4350) [inline cardano-cli/test/cardano-cli-golden/Test/Golden/ErrorsSpec.hs]: Left over comment?

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs:129]: It would be a good exercise for you to have a look in [https://github.com/input-output-hk/cardano-ledger](https://github.com/input-output-hk/cardano-ledger) to see if there are any restrictions on the drep metadata. Then double check that we are either:
- Using functionality from cardano-ledger that automatically enforces these restrictions
or
- Implement our own functionality in `cardano-api` to enforce these restrictions

**Review by Jimbo4350 — COMMENTED** [review]
LGTM, just a couple of comments. I can approve after they are addressed.

File: cardano-cli/test/cardano-cli-golden/Test/Golden/ErrorsSpec.hs
Comment (carlhammann) [reply cardano-cli/test/cardano-cli-golden/Test/Golden/ErrorsSpec.hs]: Yes. I removed it now, but it was there to remind myself (as I wrote [above](https://github.com/input-output-hk/cardano-cli/pull/308#issuecomment-1735368480))

> We can't test the GovernanceCmdDRepMetadataValidationError is redered correctly, at least not without a hack, since the constructor of DRepMetadataValidationError is not exported by cardano-api, it seems. Does this warrant a PR there?

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs
Comment (carlhammann) [reply cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs:129]: Thanks for the advice! I am using `validateAndHashDRepMetadata`, which comes from `cardano-api`, and I was assuming that, if that function exists, it is what I should use. As far as I can tell, the only thing it enforces is that the length of the metadata is at most 512 bytes. 

I can't seem to find anything in `cardano-ledger` that might be the reason for this restriction, though. The [relevant CIP](https://github.com/cardano-foundation/CIPs/blob/e96681908fd336702b9d89827063a190c5656eab/CIP-1694/README.md?plain=1#LL398C45-L398C45) says that the format of the metadata is left unspecified, and that its hash shouldn't be checked by on-chain logic (which might explain why I can't find anything in `cardano-ledger`?), but that client applications should perform "the usual sanity checks"... Is the 512-byte restriction an "usual sanity check" in this context?

**Comment by carlhammann** [comment]
I think I addressed everything mentioned in the discussions. Shall I press "merge", @Jimbo4350?

**Comment by CarlosLopezDeLara** [comment]
In terms of functionality it is good to merge, thank you!

```
➜  b2sum -l 256 drep-metadata.json
db6b415cd29ed231d6a2c41e309c6d8e2d0b9fab84c47db8c4c43ff1e055d93c  drep-metadata.json
➜  cardano-cli conway governance drep metadata-hash --drep-metadata-file drep-metadata.json
db6b415cd29ed231d6a2c41e309c6d8e2d0b9fab84c47db8c4c43ff1e055d93c
```

**Comment by carlhammann** [comment]
@CarlosLopezDeLara Thanks! It think it makes sense to add this as a test, so I did: [c4c8ef6](https://github.com/input-output-hk/cardano-cli/pull/308/commits/c4c8ef6c7dc513f419c6722d52bb937c8bb5acb5).

**Comment by carlhammann** [comment]
... and now CI is failing on Windows, where the computed hash is different (?). Does that ring a bell to you @CarlosLopezDeLara?

File: cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs
Comment (CarlosLopezDeLara) [reply cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/DRep.hs:129]: I think this 512-byte rule actually comes from the stake pool metadata on the shelley spec.  It is safe use this same rule for DReps until there is  a CIP approved for drep metadata.

**Comment by smelc** [comment]
> and now CI is failing on Windows, where the computed hash is different (?). Does that ring a bell to you

@carlhammann> The CI is a bit flaky :sweat: , and in this case, you need to remove the cabal cache as follows:

https://github.com/input-output-hk/cardano-node/pull/5456/files#diff-c09cc8dc973415b1221308559b96942a1c4b872e690bf360fc22d9b42822d519

**Pull request merged**
